Refactor API exposure tutorials (#15332)
kyma-project/kyma: docs/03-tutorials/00-api-exposure/apix-02-setup-custom-domain-for-workload.md

---
title: Set up a custom domain for a workload
---

This tutorial shows how to set up your custom domain and prepare a certificate for exposing a workload. It uses Gardener [External DNS Management](https://github.com/gardener/external-dns-management) and [Certificate Management](https://github.com/gardener/cert-management) components.
  >**NOTE:** Feel free to skip this tutorial if you use a Kyma domain instead of your custom domain.

## Prerequisites

To follow this tutorial, use Kyma 2.0 or higher.

If you use a cluster not managed by Gardener, install the [External DNS Management](https://github.com/gardener/external-dns-management) and [Certificate Management](https://github.com/gardener/cert-management) components manually in a dedicated Namespace.

This tutorial is based on a sample HttpBin service deployment and a sample Function. To deploy or create those workloads, follow the [Create a workload](./apix-01-create-workload.md) tutorial.

## Steps

Follow these steps to set up your custom domain and prepare a certificate required to expose a workload.

1. Create a Secret containing credentials for your DNS cloud service provider account in your Namespace.

>   See the [official External DNS Management documentation](https://github.com/gardener/external-dns-management/blob/master/README.md#external-dns-management), choose your DNS cloud service provider, and follow the relevant guidelines. Then run:

Reviewer: ```suggestion
  See the [official External DNS Management documentation](https://github.com/gardener/external-dns-management/blob/master/README.md#external-dns-management), choose your DNS cloud service provider, and follow the relevant guidelines to create a secret in your Namespace. Then run:
```

---
title: Set up a custom domain for a workload
---

This tutorial shows how to set up your custom domain and prepare a certificate for exposing a workload. It uses Gardener [External DNS Management](https://github.com/gardener/external-dns-management) and [Certificate Management](https://github.com/gardener/cert-management) components.
  >**NOTE:** Feel free to skip this tutorial if you use a Kyma domain instead of your custom domain.

## Prerequisites

If you use a cluster not managed by Gardener, install the [External DNS Management](https://github.com/gardener/external-dns-management) and [Certificate Management](https://github.com/gardener/cert-management) components manually in a dedicated Namespace.

This tutorial is based on a sample HttpBin service deployment and a sample Function. To deploy or create those workloads, follow the [Create a workload](./apix-01-create-workload.md) tutorial.

> 

Reviewer 2: what do you mean with sample function? the 01-create-workflow guide deploys just httpbin as described in:
https://raw.githubusercontent.com/istio/istio/master/samples/httpbin/httpbin.yaml

Reviewer 2: my bad, didn't see the Function section.

   ```bash
   export SECRET={SECRET_NAME}
   ```

2. Create a DNSProvider and a DNSEntry custom resource (CR).

   - Export the following values as environment variables and run the command provided.
  
   As the **SPEC_TYPE**, use the relevant provider type. See the [official Gardener examples](https://github.com/gardener/external-dns-management/tree/master/examples) of the DNSProvider CR.

   ```bash
   export SPEC_TYPE={PROVIDER_TYPE}
   export DOMAIN_TO_EXPOSE_WORKLOADS={DOMAIN_NAME} 
   ```
   >**NOTE:** `DOMAIN_NAME` is the domain that you own, for example, mydomain.com

   ```bash
   cat <<EOF | kubectl apply -f -
   apiVersion: dns.gardener.cloud/v1alpha1
   kind: DNSProvider
   metadata:
     name: dns-provider
     namespace: $NAMESPACE
     annotations:
       dns.gardener.cloud/class: garden
   spec:
     type: $SPEC_TYPE
     secretRef:
       name: $SECRET
     domains:
       include:
         - $DOMAIN_TO_EXPOSE_WORKLOADS
   EOF
   ```

   - Export the following values as environment variables and run the command provided:

   ```bash

>    export IP=$(kubectl -n istio-system get service istio-ingressgateway -o jsonpath='{.status.loadBalancer.ingress[0].ip}') # Assuming only one LoadBalancer with external IP

Reviewer 2: Can you please extend here the comment or add is as a note that for getting the IP:
`Sometimes you may need to use jsonpath='{.status.loadBalancer.ingress[0].hostname}' instead of ip, depending on cluster provider, for example on AWS`

---
title: Set up a custom domain for a workload
---

This tutorial shows how to set up your custom domain and prepare a certificate for exposing a workload. It uses Gardener [External DNS Management](https://github.com/gardener/external-dns-management) and [Certificate Management](https://github.com/gardener/cert-management) components.

>   >**NOTE:** Feel free to skip this tutorial if you use a Kyma domain instead of your custom domain.

Reviewer 3: ```suggestion
  >**NOTE:** Skip this tutorial if you use a Kyma domain instead of your custom domain.
```
If you don't need it let's skip it :D

   ```bash
   export SPEC_TYPE={PROVIDER_TYPE}

>    export SECRET={SECRET_NAME}

Reviewer: it can be removed since it was exported in previous step

     commonName: $DOMAIN_TO_EXPOSE_WORKLOADS
     issuerRef:
       name: $ISSUER
       namespace: $NAMESPACE
   EOF
   ```

>    >**NOTE:** Run the following command to check the certificate status: `kubectl get certificate httpbin-cert -n istio-system `

Reviewer 2: extra whitespace at end of command.

     commonName: $DOMAIN_TO_EXPOSE_WORKLOADS
     issuerRef:
       name: $ISSUER
       namespace: $NAMESPACE
   EOF
   ```

>    >**NOTE:** Run the following command to check the certificate status: `kubectl get certificate kyma-gateway-certs -n istio-system `

Reviewer: ```suggestion
   >**NOTE:** Run the following command to check the certificate status: `kubectl get certificate httpbin-cert -n istio-system `
```
